packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/MongoDBSerializer.kt

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright 2024 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.kotlin.mongodb.internal
+
+import io.realm.kotlin.internal.RealmObjectCompanion
+import io.realm.kotlin.internal.asBsonBinary
+import io.realm.kotlin.internal.asBsonDateTime
+import io.realm.kotlin.internal.asRealmInstant
+import io.realm.kotlin.internal.asRealmUUID
+import io.realm.kotlin.internal.platform.realmObjectCompanionOrThrow
+import io.realm.kotlin.internal.realmObjectCompanionOrNull
+import io.realm.kotlin.internal.util.Validation
+import io.realm.kotlin.schema.RealmProperty
+import io.realm.kotlin.schema.RealmStorageType
+import io.realm.kotlin.types.BaseRealmObject
+import io.realm.kotlin.types.RealmAny
+import io.realm.kotlin.types.RealmInstant
+import io.realm.kotlin.types.RealmObject
+import io.realm.kotlin.types.RealmUUID
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import org.mongodb.kbson.BsonBinary
+import org.mongodb.kbson.BsonBoolean
+import org.mongodb.kbson.BsonDocument
+import org.mongodb.kbson.BsonDouble
+import org.mongodb.kbson.BsonInt64
+import org.mongodb.kbson.BsonNull
+import org.mongodb.kbson.BsonString
+import org.mongodb.kbson.BsonType
+import org.mongodb.kbson.BsonValue
+import org.mongodb.kbson.Decimal128
+import org.mongodb.kbson.ObjectId
+import kotlin.reflect.KClass
+import kotlin.reflect.KMutableProperty1
+import kotlin.reflect.KProperty1
+
+/**
+ * Serializer that will encode and decode realm objects to and from EJSON sent and received
+ * from the MongoClient APIs according to the schema definition of the realm objects.
+ *
+ * Serialization of links will only include primary key of the target and deserialization of
+ * responses from MongoClient will create target link instances where only the primary key is set
+ * (and all the other properties of the object will have default values).
+ *
+ * The target types of links in mixed fields cannot be derived from the schema definition of the
+ * realm objects. To be able to deserialize and create the correct instance, the serializer needs to
+ * know of all potential target types.
+ */
+public open class MongoDBSerializer internal constructor(
```

Comment:
I think we could use some tests for this serializer.

Reply:
Added some unit tests for it in [MongoDBSerializerTests](https://github.com/realm/realm-kotlin/pull/1593/files#diff-f116755158ccc2fa732e28db602c505543f8fe34a6b94445aa9316dba405a52)